src/lib/pubkey/hss_lms/lms.h

```diff
@@ -0,0 +1,342 @@
+      std::string m_hash_name;
+};
+
+/**
+ * @brief Base class for LMS private and public key. Contains public data associated with this
+ *        LMS instance.
+ */
+class BOTAN_TEST_API LMS_Instance {
+   public:
+      /**
+       * @brief Constructor storing the provided LMS data.
+       */
+      LMS_Instance(LMS_Params lms_params, LMOTS_Params lmots_params, LMS_Identifier identifier) :
+            m_lms_params(std::move(lms_params)),
+            m_lmots_params(std::move(lmots_params)),
+            m_identifier(std::move(identifier)) {}
+
+      /**
+       * @brief The LMS parameters for this LMS instance.
+       */
+      const LMS_Params& lms_params() const { return m_lms_params; }
+
+      /**
+       * @brief The LMOTS parameters used for OTS instances of this LMS instance.
+       */
+      const LMOTS_Params& lmots_params() const { return m_lmots_params; }
+
+      /**
+       * @brief The identifier of this LMS tree ('I' in RFC 8554)
+       */
+      const LMS_Identifier& identifier() const { return m_identifier; }
+
+   private:
+      LMS_Params m_lms_params;
+      LMOTS_Params m_lmots_params;
+      LMS_Identifier m_identifier;
+};
+
+class LMS_PublicKey;
+
+/**
+ * @brief Representation of an LMS Private key
+ *
+ * Contains the secret seed used for OTS key derivation
+ * as described in RFC 8554 Appendix A.
+ */
+class BOTAN_TEST_API LMS_PrivateKey : public LMS_Instance {
+   public:
+      /**
+       * @brief Construct storing the LMS instance data and the secret seed
+       */
+      LMS_PrivateKey(LMS_Params lms_params, LMOTS_Params lmots_params, LMS_Identifier I, LMS_Seed seed) :
+            LMS_Instance(std::move(lms_params), std::move(lmots_params), std::move(I)), m_seed(std::move(seed)) {}
+
+      /**
+       * @brief The secret seed used for LMOTS' WOTS chain input creation (RFC 8554 Appendix A)
+       */
+      const LMS_Seed& seed() const { return m_seed; }
+
+      /**
+       * @brief Sign a message using an LMS_PrivateKey and the used leaf index (RFC 8554 5.4.1).
+       *
+       * The signature is written in the provided buffer. The LMS_PublicKey
+       * associated with the given private key is returned.
+       */
+      LMS_PublicKey sign_and_get_pk(StrongSpan<LMS_Signature_Bytes> out_sig,
+                                    LMS_Tree_Node_Idx q,
+                                    const LMS_Message& msg) const;
+
+   private:
+      LMS_Seed m_seed;
+};
+
+class LMS_Signature;
+
+/**
+ * @brief The LMS public key.
+ *
+ * Format according to RFC 8554:
+ * u32str(type) || u32str(otstype) || I || T[1]
+ */
+class BOTAN_TEST_API LMS_PublicKey : public LMS_Instance {
+   public:
+      /**
+       * @brief Parse a public LMS key.
+       *
+       * @param slicer The BufferSlicer at the public key bytes' position
+       * @return The LMS public key.
+       * @throws Decoding_Error If parsing the public key fails.
+       */
+      static LMS_PublicKey from_bytes_or_throw(BufferSlicer& slicer);
```

Comment:
I agree that we can pass a span here instead. This would be suitable but inconsistent with signature parsing, where passing the slicer is quite sensible (see below). I have no hard feelings for both options.

Reply:
Perhaps my two cents; talking from experience in #4024: I totally agree, that any loosely coupled API shouldn't pass around `BufferSlicer` or `BufferSlicer` references. Though, for tightly coupled internal APIs (e.g. utility functions in the same compilation unit), I find this very much convenient.

Accepting a span, as @randombit suggested, has the drawback that _the caller must know how many bytes_ the callee will want to consume or emit. Usually, that can be determined, but it typically requires some extra calculations that were already done to determine the overall size of the final serialized structure. Just passing the `Buffer***` reference and trusting that the callee will be responsible, is saving some cycles and also noise in the code.

That said: We should always remember to validate that the `Buffer***` is full/empty when expected at the very end of the marshalling, to check that everything checked out as expected.